src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -314,15 +315,123 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
                                                            chip::NodeId CaseAdminNode, uint16_t AdminVendorId)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_FAILURE;
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> cert;
+
+    uint8_t * certBuf = nullptr;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has added an Op Cert");
+
+    AdminPairingInfo * admin = retrieveCurrentAdmin();
+    VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
+
+    VerifyOrExit(cert.Alloc(NOC.size() + ICACertificate.size()), status = EMBER_ZCL_STATUS_FAILURE);
+    certBuf = cert.Get();
+    memcpy(certBuf, NOC.data(), NOC.size());
+    memcpy(&certBuf[NOC.size()], ICACertificate.data(), ICACertificate.size());
+
+    VerifyOrExit(admin->SetOperationalCert(ByteSpan(certBuf, NOC.size() + ICACertificate.size())) == CHIP_NO_ERROR,
+                 status = EMBER_ZCL_STATUS_FAILURE);
+
+exit:
     emberAfSendImmediateDefaultResponse(status);
+    if (status == EMBER_ZCL_STATUS_FAILURE)
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Failed AddOpCert request.");
+    }
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to encode response command: %s", ErrorStr(err));
+    }
+
     return true;
 }
 
 bool emberAfOperationalCredentialsClusterOpCSRRequestCallback(chip::app::Command * commandObj, chip::ByteSpan CSRNonce)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_FAILURE;
-    emberAfSendImmediateDefaultResponse(status);
+    EmberAfStatus status   = EMBER_ZCL_STATUS_SUCCESS;
+    EmberStatus sendStatus = EMBER_SUCCESS;
+    CHIP_ERROR err         = CHIP_NO_ERROR;
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> csr;
+    size_t csrLength = Crypto::kMAX_CSR_Length;
+
+    uint8_t * csrBuf = nullptr;
+
+    constexpr uint16_t zclOctetStringHeaderSize = 1;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has requested an OpCSR");
+
+    // Fetch current admin
+    AdminPairingInfo * admin = retrieveCurrentAdmin();
+    VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
+
+    // 1 extra byte for length in ZCL data type
+    VerifyOrExit(csr.Alloc(Crypto::kMAX_CSR_Length + zclOctetStringHeaderSize), status = EMBER_ZCL_STATUS_FAILURE);
+
+    if (admin->GetOperationalKey() == nullptr)
+    {
+        Crypto::P256Keypair keypair;
+        keypair.Initialize();
+        VerifyOrExit(admin->SetOperationalKey(keypair) == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
+    }
+
+    csrBuf = csr.Get();
+    err    = admin->GetOperationalKey()->NewCertificateSigningRequest(&csrBuf[zclOctetStringHeaderSize], csrLength);
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: NewCertificateSigningRequest returned %d", err);
+    VerifyOrExit(err == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(csrLength <= UINT8_MAX, status = EMBER_ZCL_STATUS_FAILURE);
```

Comment:
```suggestion
    VerifyOrExit(csrLength < UINT8_MAX, status = EMBER_ZCL_STATUS_FAILURE);
```
Strings are not allowed to have length 255, right?